### PR TITLE
util/cloudenv: add Azure support & DNS IPs

### PR DESCRIPTION
**Pull request by bradfitz**
And rewrite cloud detection to try to do only zero or one metadata
discovery request for all clouds, only doing a first (or second) as
confidence increases. Work remains for Windows, but a start.

And add Cloud to tailcfg.Hostinfo, which helped with testing using
"tailcfg debug hostinfo".

Updates #4983 (Linux only)
Updates #4984

### CODE REVIEW

File: util/cloudenv/cloudenv.go [hunk 1]
```diff
@@ -66,20 +81,88 @@ func Get() Cloud {
...
+	tr := &http.Transport{
+		DisableKeepAlives: true,
+		Dial: (&net.Dialer{
+			Timeout: maxWait,
+		}).Dial,
+	}
+	ctx, cancel := context.WithTimeout(context.Background(), maxWait)
+	defer cancel()
+
+	req, _ := http.NewRequestWithContext(ctx, "HEAD", "http://"+CommonNonRoutableMetadataIP, nil)
```
Comment (DentonGentry) [inline util/cloudenv/cloudenv.go:134]: I think this turns into a IMDSv1 request, a GET (HEAD, in this case) to the local Metadata server. IMDSv1 is disabled in many AWS deployments now and, more problematically, may have CloudWatch configured to watch for IMDSv1 requests.

Comment (bradfitz) [reply util/cloudenv/cloudenv.go:134]: AWS IMDSv1 and AWS IMDSv2 and GCP metadata and Azure metadata are all four just HTTP to this same IP address.

To use IMDSv2 you need to make an HTTP request like this without any token to get a token. (`curl -X PUT "http://169.254.169.254/latest/api/token" -H "X-aws-ec2-metadata-token-ttl-seconds: 60"`)

The CloudWatch `MetadataNoToken` metric is mentioned at https://docs.aws.amazon.com/AWSEC2/latest/UserGuide/viewing_metrics_with_cloudwatch.html but doesn't say what exactly it counts. I confidently assume it doesn't count requests to get a token. I less confidently assume it shouldn't count no-op requests like a `HEAD /` that aren't trying to do anything, but who knows.

I could try to get a token instead, but that might just fire a CloudWatch-like counter on Azure or GCP instead.

I suppose I could experiment with CloudWatch and see if it triggers with this.

Looking at their SDK,

https://docs.aws.amazon.com/sdk-for-go/api/aws/ec2metadata/#EC2Metadata.Available
```
// Available returns if the application has access to the EC2 Metadata service.
// Can be used to determine if application is running within an EC2 Instance and
// the metadata service is available.
func (c *EC2Metadata) Available() bool {
```

.... it fetches a token (which would've been enough to answer the question of whether EC2Metadata was available) and then proceeds to fetch the instance-id using that token.

So we could fetch a token I guess?

But look a few lines up: on Linux, we don't even make it this far if we already found from the /sys info that we're on EC2. So this path would only trigger for Windows in the future (currently Windows returns false for all clouds)

Comment (bradfitz) [reply util/cloudenv/cloudenv.go:134]: Well, ....

```
$ for x in $(seq 0 500); do curl  -I http://169.254.169.254/; done
....
```
Screen Shot 2022-06-30 at 8 50 06 AM

That answers that question.

Comment (bradfitz) [reply util/cloudenv/cloudenv.go:134]: But no spike from:

```
ubuntu@bradprod:~$ for x in $(seq 0 500); do curl -X PUT "http://169.254.169.254/latest/api/token" -H "X-aws-ec2-metadata-token-ttl-seconds: 5" ; done
```

Screen Shot 2022-06-30 at 9 23 55 AM

Comment (bradfitz) [reply util/cloudenv/cloudenv.go]: @DentonGentry, PTAL.

File: util/cloudenv/cloudenv.go [hunk 2]
```diff
@@ -66,20 +82,102 @@ func Get() Cloud {
...
+		Dial: (&net.Dialer{
+			Timeout: maxWait,
+		}).Dial,
 	}
-	if gcpmetadata.OnGCE() {
+	ctx, cancel := context.WithTimeout(context.Background(), maxWait)
+	defer cancel()
+
+	// We want to hit CommonNonRoutableMetadataIP to see if we're on AWS, GCP,
+	// or Azure. All three (and many othersA) use the same metadata IP.
```
Comment (bradfitz) [inline util/cloudenv/cloudenv.go:136]: whoops, typo. Fixed. Will push/rebase before merge.

File: util/cloudenv/cloudenv.go [hunk 3]
```diff
@@ -66,20 +82,103 @@ func Get() Cloud {
...
+	defer cancel()
+
+	// We want to hit CommonNonRoutableMetadataIP to see if we're on AWS, GCP,
+	// or Azure. All three (and many others) use the same metadata IP.
+	//
+	// But to avoid triggering the AWS CloudWatch "MetadataNoToken" metric (for which
+	// there might be an alert registered?), make our initial request be a token
+	// request. This only works on AWS, but the failing HTTP response on other clouds gives
+	// us enough clues about which cloud we're on.
+	req, err := http.NewRequestWithContext(ctx, "PUT", "http://"+CommonNonRoutableMetadataIP+"/latest/api/token", strings.NewReader(""))
```
Comment (DentonGentry) [inline util/cloudenv/cloudenv.go:142]: To my knowledge, a PUT will be ok.

**Review by DentonGentry — APPROVED** [review]

**Pull request merged**